operators/pkg/custom-metrics-server/metrics_scraper.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2020-2022 Politecnico di Torino
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package custom_metrics_server
+
+import (
+	context "context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+
+	"github.com/netgroup-polito/CrownLabs/operators/pkg/utils"
+)
+
+// CustomMetrics stores desired metrics of a container.
+type CustomMetrics struct {
+	CPUPerc   float32 `json:"cpu"`
+	MemBytes  uint64  `json:"mem"`
+	DiskBytes uint64  `json:"disk"`
+}
+
+func (c CustomMetrics) String() string {
+	return fmt.Sprintf("[CPU] %v \t[MEM] %v \t[DISK] %v", c.CPUPerc, c.MemBytes, c.DiskBytes)
+}
+
+var (
+	// Containers ignored wile retrieving metrics.
+	ignoredContainerNames = []string{"xvnc", "websockify", "content-downloader", "custom-metrics-server-daemon"}
+	// List of watched containerIds.
+	containerIds = []string{}
+	// Stats used to evaluate resource usages changes
+	// <key=podName, val=ContainerStats>.
+	oldStats = make(map[string]*runtimeapi.ContainerStats)
+	// Stats used to evaluate resource usages changes
+	// <key=podName, val=CustomMetrics>.
+	cachedMetrics      = make(map[string]*CustomMetrics)
+	containerIdsMutex  = sync.RWMutex{}
+	cachedMetricsMutex = sync.RWMutex{}
+)
+
+func runMetricsScraper(log logr.Logger, updatePeriod time.Duration) {
+	// Initlize containerIds list.
+	if err := findApplicationContainerIds(log.WithName("find-application-container-ids").
+		WithValues("reason", "initialize containerIds list"),
+		5*time.Second); err != nil {
+		log.Error(err, "Error initializing containerIds list")
+	}
+
+	// Periodically update containerIds list.
+	go updateApplicationContainerIds(log.WithName("update-application-container-ids"))
```

Comment:
Would merge the update of the container IDs and the retrieval of the metrics, as it doesn't bring any advantage to make them asynchronous.

Reply:
Merged the two goroutines. They were designed separately to define different frequencies of metrics and running containers retrieval. 
While metrics are usually required to be updated frequently, watching for new running container may have been a less frequent task